### PR TITLE
implements ArrayList with method add(), add(index), addAll(), get(), …

### PR DESCRIPTION
**Pull request by mkarmadev**
…set(), remove(index), remove(), size(), isEmpty(), checkIndex(), checkIndexForAdd(), growIfArrayFull()

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,101 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static int DEFAULT_CAPACITY = 10;
+    private int size = 0;
+    private Object[] storage = new Object[DEFAULT_CAPACITY];
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:9]: see common mistakes

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,101 @@
...
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index:" + index);
+        }
+    }
+
+    private void growIfArrayFull() {
+        if (size >= storage.length) {
+            storage = Arrays.copyOf(storage, storage.length + (storage.length >> 1));
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:98]: use System.arrayCopy()

**Review by aaananas — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,106 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static int DEFAULT_CAPACITY = 10;
+    private int size = 0;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:7]:
```suggestion
    private int size;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,106 @@
...
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (storage[i] == element || (storage[i] != null && storage[i].equals(element))) {
+                return remove(i);
+            }
+        }
+
+        throw new NoSuchElementException("Not found element");
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:71]:
```suggestion
        throw new NoSuchElementException("Not found element " + element );
```

**Review by Rommelua — APPROVED** [review]

**Pull request closed without merging**